### PR TITLE
[Transform] Transform optmize date histogram

### LINKED ISSUE #54254: [Transform] continuous transform date_histogram group_by performance
**Affected versions**: <= 7.6

**Problem** 

Continuous Transforms are optimized for usecases, where sessions are grouped using [terms](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-bucket-terms-aggregation.html). Grouping on `date_histogram` - e.g. per hour metrics -  with large datasets suffers from re-writing _all_ buckets for every checkpoint. This causes a lot of load on the cluster and might result in service degradation.

**Mitigation** 

[Rollup](https://www.elastic.co/guide/en/elasticsearch/reference/master/rollup-overview.html) is optimized for this usecase and provides - via rollup search - aggregations on aggregations. Please consider using rollup instead.

Transform will provide an optimization for grouping on date_histogram with version **TBD**. Please consider upgrading to **TBD**. (Note that you can use a separate cluster for transform as transform supports CCS)

For this optimization to kick in, the field you configure for `sync` _must_ be the same field you configure for the date_histogram `group_by`. Using multiple `group_by` is still possible, the transform gets optimized for the group_by where the field matches the field for time-based `sync`.

### PR DESCRIPTION
**Pull request by hendrikmuhs**
optimize transform for group_by on date_histogram by injecting an additional range query. This limits the number of search and index requests and avoids unnecessary updates. Only recent buckets get re-written.

fixes #54254

### CODE REVIEW

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/pivot/Pivot.java [hunk 1]
```diff
@@ -189,44 +189,37 @@ public boolean supportsIncrementalBucketUpdate() {
...
-            if (changedBuckets.containsKey(entry.getKey())) {
-                return entry.getValue().getIncrementalBucketUpdateFilterQuery(changedBuckets.get(entry.getKey()));
-            } else {
-                // should never happen
-                throw new RuntimeException("Could not find bucket value for key " + entry.getKey());
-            }
+            Set<String> changedBucketsByGroup = changedBuckets.get(entry.getKey());
+
+            // important: the fields must match to apply this optimization
+            long synchronizationTimestamp = entry.getKey().equals(synchronizationField) ? lastSynchronizationCheckpoint : 0;
```
Comment (benwtrent) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/pivot/Pivot.java]: It seems we are needlessly adding an if branch. I think we should probably just supply both the `synchronizationField` and the `lastSynchronizationCheckpoint` to `getIncrementalBucketUpdateFilterQuery`. 

(similar comment for the more than one group source entry)

**Review by benwtrent — COMMENTED** [review]
The optimization makes sense. If we are pivoting with a date_histogram, we shouldn't run through the entire index to update the destination.

The state machine changes were difficult to parse through. Will go through them again once this is no longer a draft.

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/pivot/Pivot.java (hunk 1 above)
Comment (hendrikmuhs) [reply x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/pivot/Pivot.java]: :+1: This is a bug, `entry.getKey()` returns the name of the group, not the field that is used (entry.getValue().getField()), but I think I will follow your suggestion and pass it in and let the GroupSource decide what to do.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/transform/transforms/pivot/DateHistogramGroupSource.java [hunk 2]
```diff
@@ -315,9 +345,13 @@ public int hashCode() {
     }
 
     @Override
-    public QueryBuilder getIncrementalBucketUpdateFilterQuery(Set<String> changedBuckets) {
-        // no need for an extra range filter as this is already done by checkpoints
-        return null;
+    public QueryBuilder getIncrementalBucketUpdateFilterQuery(Set<String> changedBuckets, long synchronizationTimestamp) {
+        if (synchronizationTimestamp > 0) {
+
+            return new RangeQueryBuilder(field).gte(rounding.round(synchronizationTimestamp)).format("epoch_millis");
```
[the remark this answers is not in the document]
Comment (hendrikmuhs) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/transform/transforms/pivot/DateHistogramGroupSource.java]: sorry, forgot to answer: yes it does, this method is used for all histogram aggregations. For getting the next bucket there is `nextRoundingValue`.

**Comment by hendrikmuhs** [comment]
run elasticsearch-ci/2

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/transform/transforms/pivot/DateHistogramGroupSource.java [hunk 3]
```diff
@@ -315,9 +344,16 @@ public int hashCode() {
...
     @Override
-    public QueryBuilder getIncrementalBucketUpdateFilterQuery(Set<String> changedBuckets) {
-        // no need for an extra range filter as this is already done by checkpoints
-        return null;
+    public QueryBuilder getIncrementalBucketUpdateFilterQuery(
+        Set<String> changedBuckets,
+        String synchronizationField,
+        long synchronizationTimestamp
+    ) {
+        if (synchronizationField != null && field != null && synchronizationField.equals(field) && synchronizationTimestamp > 0) {
```
Comment (benwtrent) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/transform/transforms/pivot/DateHistogramGroupSource.java]:
```suggestion
        if (synchronizationField != null && synchronizationField.equals(field) && synchronizationTimestamp > 0) {
```

File: x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/pivot/Pivot.java [hunk 4]
```diff
@@ -189,44 +188,31 @@ public boolean supportsIncrementalBucketUpdate() {
...
+        long lastSynchronizationCheckpoint
+    ) {
+        assert changedBuckets != null;
 
         if (config.getGroupConfig().getGroups().size() == 1) {
             Entry<String, SingleGroupSource> entry = config.getGroupConfig().getGroups().entrySet().iterator().next();
-            // it should not be possible to get into this code path
-            assert (entry.getValue().supportsIncrementalBucketUpdate());
-
             logger.trace("filter by bucket: " + entry.getKey() + "/" + entry.getValue().getField());
```
Comment (benwtrent) [inline x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/pivot/Pivot.java]:
```suggestion
            logger.trace(() -> new ParameterizedMessage("filter by bucket: {}/{}", entry.getKey(), entry.getValue().getField()));
```
nit

**Review by benwtrent — APPROVED** [review]
Another pair of 👀 would be good

**Pull request merged**